src/optimizer/filter_combiner.cpp

```diff
@@ -705,8 +705,27 @@ TableFilterSet FilterCombiner::GenerateTableScanFilters(const vector<ColumnIndex
 					}
 					auto comparison_type =
 					    invert ? FlipComparisonExpression(comp.GetExpressionType()) : comp.GetExpressionType();
-					auto const_filter = make_uniq<ConstantFilter>(comparison_type, const_val->value);
-					conj_filter->child_filters.push_back(std::move(const_filter));
+					if (const_val->value.IsNull()) {
+						switch (comparison_type) {
+						case ExpressionType::COMPARE_EQUAL:
+						case ExpressionType::COMPARE_DISTINCT_FROM: {
+							auto null_filter = make_uniq<IsNullFilter>();
+							conj_filter->child_filters.push_back(std::move(null_filter));
```

Comment:
`x = NULL` always returns `NULL` (i.e. false) -  so this should be only `DISTINCT FROM`

Reply:
Should this mean that the table scan should just be replaced with an empty result? Since `= NULL` evaluates to NULL and as a where condition this means no rows are returned. Although for nested types NULL comparisons get tricky, but we don't support NULL pushdown into nested types yet

Reply:
Yes, that's also a possible optimization, it is indeed always false

Reply:
Turns out Postgres/sqlite both evaluate `WHERE NULL or TRUE` to TRUE, so `(a = NULL OR TRUE)` should not be replaced with an empty result.

Reply:
Yes, that's SQL standard boolean logic (similarly `FALSE AND NULL` also returns `FALSE`).